### PR TITLE
build-environment: OS-specific environment keys

### PR DESCRIPTION
**Pull request by maxim-belkin**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

CC @sjackman 

Provide a way to deal with OS-specific build environment variables. On Linux, those are: LD_LIBRARY_PATH, LD_RUN_PATH, LD_PRELOAD, LIBRARY_PATH.

### CODE REVIEW

**Comment by sjackman** [comment]
I'd be inclined to add `LD_LIBRARY_PATH LD_RUN_PATH LD_PRELOAD LIBRARY_PATH` to `build_env_keys on both OS. The variables that don't affect macOS, well, won't affect it, and so do no harm. Note that `LD_RUN_PATH` and `LIBRARY_PATH` affect `gcc` even on macOS. `LD_LIBRARY_PATH` and `LD_PRELOAD` are specific to Linux's dynamic linker. On macOS these variables are called `DYLD_LIBRARY_PATH` and `DYLD_INSERT_LIBRARIES`.

**Comment by maxim-belkin** [comment]
ok, then it's up to @MikeMcQuaid: shall we add the aforementioned variables to `build_env_keys` or go through the approach implemented here?

**Comment by maxim-belkin** [comment]
Just in case, I'm ready to submit another PR: https://github.com/maxim-belkin/brew-macOS/blob/build-env2/Library/Homebrew/build_environment.rb#L49

File: Library/Homebrew/build_environment.rb [hunk 1]
```diff
@@ -46,7 +46,13 @@ def build_env_keys(env)
       HOMEBREW_SDKROOT HOMEBREW_BUILD_FROM_SOURCE
       MAKE GIT CPP
       ACLOCAL_PATH PATH CPATH
-    ].select { |key| env.key?(key) }
+    ]
+  end
+
+  def os_specific_build_env_keys; end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/build_environment.rb]: This should return an empty array or the code below won't work in the generic case.

**Comment by MikeMcQuaid** [comment]
I'm good with this approach. Can you elaborate a bit more on why this is needed, though, and what it does? Is this environment variables Homebrew/brew or formulae or users set? If the latter: when do they need to set these?

File: Library/Homebrew/build_environment.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/build_environment.rb]: `brew tests --generic` reveals this error. It's probably worth always running locally for cross-platform changes.

**Comment by sjackman** [comment]
As far as I can tell `build_env_keys` is only used by `dump_build_env` and `brew --env`. So it's for the purpose of displaying environment variables that the user has set in their environment. In a coming PR, these variables are added to `SANITIZED_VARS` of `extend/ENV/shared.rb` to ensure these environment variables don't affect `brew install`.

**Comment by sjackman** [comment]
> I'm good with this approach.

Do you prefer this PR as is or https://github.com/maxim-belkin/brew-macOS/blob/build-env2/Library/Homebrew/build_environment.rb#L49
My preference is the latter. Whichever you prefer is fine.

**Comment by MikeMcQuaid** [comment]
> As far as I can tell build_env_keys is only used by dump_build_env and brew --env. So it's for the purpose of displaying environment variables that the user has set in their environment. In a coming PR, these variables are added to SANITIZED_VARS of extend/ENV/shared.rb to ensure these environment variables don't affect brew install.

Thanks! Presumably those variables are only output if set? If so, I'm fine with just whitelisting them all without `os/extend` as you suggest. If not, to avoid extra empty output on macOS I either think we should keep this PR's approach or not output them if they aren't set.

> In a coming PR, these variables are added to SANITIZED_VARS of extend/ENV/shared.rb to ensure these environment variables don't affect brew install.

Cool. If those variables are being removed rather than added/modified I'm again 👍 on keeping that the same on macOS

**Comment by maxim-belkin** [comment]
Closing in favor of #3358.

**Pull request closed without merging**